### DISCUSSION TITLE
Heroku/Dokku deploying problems

### DISCUSSION THREAD

**Discussion by bitkidd** [adonisjs/core#2630] [Help] [answered]
Hey!
I've read several threads down here about the problem with Heroku deployments but non of the ideas there helped.
Then I tried to deploy a freshly generated app, just added a `Procfile`:
```
web: node build/server.js
```

It even build fine:
```
remote: Compressing source files... done.
remote: Building source:
remote: 
remote: -----> Building on the Heroku-20 stack
remote: -----> Using buildpack: heroku/nodejs
remote: -----> Node.js app detected
remote:        
remote: -----> Creating runtime environment
remote:        
remote:        NPM_CONFIG_LOGLEVEL=error
remote:        NODE_VERBOSE=false
remote:        NODE_ENV=production
remote:        NODE_MODULES_CACHE=true
remote:        
remote: -----> Installing binaries
remote:        engines.node (package.json):  15.x
remote:        engines.npm (package.json):   unspecified (use default)
remote:        engines.yarn (package.json):  1.x
remote:        
remote:        Resolving node version 15.x...
remote:        Downloading and installing node 15.14.0...
remote:        Using default npm version: 7.7.6
remote:        Resolving yarn version 1.x...
remote:        Downloading and installing yarn (1.22.10)
remote:        Installed yarn 1.22.10
remote:        
remote: -----> Restoring cache
remote:        Cached directories were not restored due to a change in version of node, npm, yarn or stack
remote:        Module installation may take longer for this build
remote:        
remote: -----> Installing dependencies
remote:        Installing node modules (package.json)
remote:        
remote:        added 626 packages, and audited 627 packages in 30s
remote:        
remote:        62 packages are looking for funding
remote:          run `npm fund` for details
remote:        
remote:        found 0 vulnerabilities
remote:        
remote: -----> Build
remote:        Running build
remote:        
remote:        > hello-adonis@1.0.0 build
remote:        > node ace build --production
remote:        
remote:        [ info ]  cleaning up "./build" directory
remote:        [ info ]  compiling typescript source files
remote:        [ info ]  copy { ace,package.json,package-lock.json => build }
remote:        [ info ]  copy { .adonisrc.json => build }
remote:        CREATE: ace-manifest.json file
remote:        [ success ]  built successfully
remote:        
remote:        ╭────────────────────────────────────────────────────────────────────────╮
remote:        │    Run the following commands to start the server in production        │
remote:        │────────────────────────────────────────────────────────────────────────│
remote:        │                                                                        │
remote:        │    ❯ cd build                                                          │
remote:        │    ❯ npm ci --production                                               │
remote:        │    ❯ node server.js                                                    │
remote:        │                                                                        │
remote:        ╰────────────────────────────────────────────────────────────────────────╯
remote:        
remote: -----> Caching build
remote:        - node_modules
remote:        
remote: -----> Pruning devDependencies
remote:        
remote:        up to date, audited 217 packages in 3s
remote:        
remote:        23 packages are looking for funding
remote:          run `npm fund` for details
remote:        
remote:        found 0 vulnerabilities
remote:        
remote: -----> Build succeeded!
remote: -----> Discovering process types
remote:        Procfile declares types -> web
remote: 
remote: -----> Compressing...
remote:        Done: 36.3M
remote: -----> Launching...
remote:        Released v8
remote:        https://desolate-meadow-24609.herokuapp.com/ deployed to Heroku
```

But as soon as you try to access the address, you'll get a crash, very non-descriptive:
```
2021-05-17T12:25:39.000000+00:00 app[api]: Build succeeded
2021-05-17T12:25:39.003233+00:00 heroku[web.1]: Starting process with command `node build/server.js`
2021-05-17T12:25:43.333391+00:00 heroku[web.1]: Process exited with status 0
2021-05-17T12:25:43.428844+00:00 heroku[web.1]: State changed from starting to crashed
2021-05-17T12:25:44.691809+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=desolate-meadow-24609.herokuapp.com request_id=0717b83a-851d-4ee0-8a0b-b8783ea1542e fwd="188.138.205.65" dyno= connect= service= status=503 bytes= protocol=https
```

Maybe some ideas will arise, because it feels very weird?
Thank you!

**Comment by thetutlage** [comment]
Can you share the heroku logs?

**Reply by bitkidd** [reply]
That's what `heroku logs --tail` shows

```
2021-05-17T12:25:39.000000+00:00 app[api]: Build succeeded
2021-05-17T12:25:39.003233+00:00 heroku[web.1]: Starting process with command `node build/server.js`
2021-05-17T12:25:43.333391+00:00 heroku[web.1]: Process exited with status 0
2021-05-17T12:25:43.428844+00:00 heroku[web.1]: State changed from starting to crashed
2021-05-17T12:25:44.691809+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=desolate-meadow-24609.herokuapp.com request_id=0717b83a-851d-4ee0-8a0b-b8783ea1542e fwd="188.138.205.65" dyno= connect= service= status=503 bytes= protocol=https
```

**Reply by thetutlage** [reply]
It looks like the process dies with exit code 0 after 4 seconds of running the build/server.js file.

Are these the only logs when you run `heroku logs --tail`

**Reply by bitkidd** [reply]
Yes, that's the strangest part, the process dies without notice. 
At first I was thinking that it may be somehow connected to npm, or node version, or maybe something wrong with the env values. 
But even locally, after `npm run build` and via a command `HOST=test APP_KEY=kshdfksuhdskuhdfksuhdf APP_NAME=test NODE_ENV=production PORT=1234 node build/server.js` it does nothing.

**Reply by thetutlage** [reply]
You mean it doesn't work on local either? 

Also did you check this cookbook https://docs.adonisjs.com/cookbooks/deploy-to-heroku

**Reply by bitkidd** [reply]
Yes, I have read it several times 😄 
I have just repeated the same process again:
1. `npm init adonis-ts-app@latest hello-world`
2. Created `Procfile`
3. Initiated `git`
4. Created a new Heroku app
5. Added `APP_KEY` and `HOST` env variables
6. Deployed 
7. Tried to access, no luck

The build itself passes, container just can't start.

And even locally, after build, server does not start if I run:
`HOST=test PORT=1234 NODE_ENV=production APP_KEY=skjhdf23sdfsk3453dsdfhfsdk node build/server.js`

**Reply by thetutlage** [reply]
If its not working locally, then I can give it a try as well.

You say the container doesn't work. Are you running inside docker container? If yes, can you share the dockerfile?

**Reply by bitkidd** [reply]
For this I have't used docker, just ran those commands and tried to understand what is the reason for it not to start.
My assumption about containers not being able to start is based on heroku reports where they just crashed for some reason.

**Reply by thetutlage** [reply]
Tried it myself on local. Yes it did fail, but I get the error too

<img width="939" alt="Screen Shot 2021-05-17 at 11 07 02 PM" src="https://user-images.githubusercontent.com/1706381/118532505-a94de880-b764-11eb-8ac5-c19a5281e11b.png">

**Reply by bitkidd** [reply]
Yes, that kind of error I've seen too, did it start as you add this variable? For me it didn't work even after I added it.

**Reply by thetutlage** [reply]
Yes it does. 

- This is a fresh AdonisJS application
- I have not added or changed anything
- Built the app using `node ace build --production`
- Not running inside any container. Directly on the host machine

<img width="1017" alt="Screen Shot 2021-05-18 at 12 10 34 AM" src="https://user-images.githubusercontent.com/1706381/118539815-83791180-b76d-11eb-8b14-e247e128227e.png">

**Reply by bitkidd** [reply]
Hm. Not sure what it may be, I just removed node and volta completely, installed both again, initiated `adonis slim`, built it and could not run using `node build/server.js` with `ENV` variables. It shows the error when variables are not there, but as soon as I add them, it just does not start.

The strangest part here is that other adonis v5 apps start and work, just a newly created one does not 🤔 😫

**Reply by thetutlage** [reply]
Can you publish the new app to a repo and share the link here?

**Reply by bitkidd** [reply]
Sure. Here it is: https://github.com/bitkidd/hello-world

**Reply by thetutlage** [reply]
Still works like a charm for me. Have you tried any different ports and can try to run with `LOG_LEVEL=trace` environment variable and share the output?

<img width="872" alt="Screen Shot 2021-05-18 at 12 28 46 PM" src="https://user-images.githubusercontent.com/1706381/118606172-ba364280-b7d4-11eb-82ca-a7a32ecf8718.png">

**Reply by bitkidd** [reply]
Just ran it with tracing and tried several different ports:
```bash
➜  build git:(main) HOST=testtest APP_KEY=298742342bnkjhdi7w8yiehw9ei8yhek PORT=5678 NODE_ENV=production LOG_LEVEL=trace node server.js
{"level":10,"time":1621321632237,"pid":94355,"hostname":"MacBook-Pro.local","name":"hello-world","msg":"registering providers"}
{"level":10,"time":1621321632242,"pid":94355,"hostname":"MacBook-Pro.local","name":"hello-world","msg":"booting providers"}
{"level":10,"time":1621321632287,"pid":94355,"hostname":"MacBook-Pro.local","name":"hello-world","file":"./start/routes","optional":false,"environment":["web","console","test","repl"],"msg":"file:preload"}
{"level":10,"time":1621321632287,"pid":94355,"hostname":"MacBook-Pro.local","name":"hello-world","file":"./start/kernel","optional":false,"environment":["web","console","test","repl"],"msg":"file:preload"}
{"level":10,"time":1621321632288,"pid":94355,"hostname":"MacBook-Pro.local","name":"hello-world","msg":"optimizing http server handler"}
{"level":10,"time":1621321632290,"pid":94355,"hostname":"MacBook-Pro.local","name":"hello-world","msg":"binding App/Exceptions/Handler exception handler"}
{"level":10,"time":1621321632291,"pid":94355,"hostname":"MacBook-Pro.local","name":"hello-world","msg":"executing providers ready hook"}
{"level":10,"time":1621321632297,"pid":94355,"hostname":"MacBook-Pro.local","name":"hello-world","msg":"forcefully killing http server"}
{"level":10,"time":1621321632298,"pid":94355,"hostname":"MacBook-Pro.local","name":"hello-world","msg":"closing http server"}
{"level":10,"time":1621321632300,"pid":94355,"hostname":"MacBook-Pro.local","name":"hello-world","msg":"executing providers shutdown hook"}
```

**Reply by thetutlage** [reply]
`forcefully killing http server` is logged by the `kill` method https://github.com/adonisjs/core/blob/c3a29e8ef17d947e7e5653205b702c40faacd2e6/src/Ignitor/HttpServer/index.ts#L209. Can you open the `node_modules` and manually drop a `debugger` statement to see when and from this method is called?

**Reply by bitkidd** [reply]
Just did it:
```
Uncaught /hello-world/node_modules/@poppinss/utils/build/src/Helpers/string.js:241
```

```
/hello-world/node_modules/@poppinss/utils/build/src/Helpers/string.js:241
        return `${values[0]}${options?.pairSeparator || ' and '}${values[1]}`;
                                      ^

SyntaxError: Unexpected token '.'
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/hello-world/node_modules/@poppinss/utils/build/src/Helpers/index.js:56:31)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
```

**Reply by thetutlage** [reply]
What version of Node.js you are running?

**Reply by bitkidd** [reply]
node v14.17.0
npm 6.14.13

**Reply by thetutlage** [reply]
I just tried it on Node 12.0.0 and does report the error to me.

<img width="1339" alt="Screen Shot 2021-05-18 at 12 58 48 PM" src="https://user-images.githubusercontent.com/1706381/118609886-d5a34c80-b7d8-11eb-83de-ebad082ec3f8.png">

**Reply by thetutlage** [reply]
Are you sure that it is `v14.17.0`? Coz, the app works fine with 14.17.0 for me. Can you log it from the node process. Just write the following line inside `build/server.js` file

```
console.log(process.version)
```

**Reply by bitkidd** [reply]
Yeah, just checked it like this, on v12 it reports an error too, on v14 nothing, on v16 nothing.

![Screenshot 2021-05-18 at 10 34 32](https://user-images.githubusercontent.com/2136460/118610630-a4208600-b7c4-11eb-8c13-9c0749a9cb7f.png)

**Reply by thetutlage** [reply]
But it is supposed to run on v14 and v16 and not have the syntax error at first place. 🤔

**Reply by bitkidd** [reply]
Yes, and the strangest thing here is that basically the same project (an older app updated from preview), with the same packages works well on v14 and v16. And it deploys well too without any problem.

**Reply by bitkidd** [reply]
Okay, I guess I found it. If I use `HOST=test` variable it will not start, but it will start with `HOST=0.0.0.0` or `HOST=localhost` or if I remove `HOST` from `ent.ts` and variables entirely.

**Reply by bitkidd** [reply]
Yes, just deployed it, so the problem was with `HOST` env. Super strange.

**Reply by bitkidd** [reply]
@thetutlage thank you for your time and sorry for bothering you with this!

**Comment by bitkidd** [comment, marked as answer]
Okay, I guess I found it. If I use `HOST=test` variable it will not start, but it will start with `HOST=0.0.0.0` or `HOST=localhost` or if I remove `HOST` from `ent.ts` and variables entirely.

**Reply by maswebber** [reply]
wow i change to 0.0.0.0 the program run, before the HOST was localhost
